File: src/main/java/grojdg/harderstats/events/ServerPlayerEntityEvents.java
package grojdg.harderstats.events;

import grojdg.harderstats.HarderStats;
import grojdg.harderstats.InfoReceptionService;
import net.minecraft.entity.damage.DamageSource;
import net.minecraft.server.network.ServerPlayerEntity;
import net.minecraft.server.world.ServerWorld;
import net.minecraft.util.ActionResult;
import net.minecraft.world.World;

public class ServerPlayerEntityEvents {
    public static ActionResult onDamage(DamageSource source, float amount, ServerPlayerEntity player) {
        if((!source.getName().equals("hih") && !source.getType().msgId().equals("outOfWorld"))) {
            InfoReceptionService.InfoReceptionServiceFactory.get().updateDamageTaken(player.getGameProfile().getId(), (int) amount);
        }

        return ActionResult.PASS;
    }

    public static ActionResult onDeath(DamageSource source, ServerPlayerEntity player) {
        // this is the type we kill with manually in HeartOfHearts
        if (!source.getType().msgId().equals("outOfWorld")) {
            String entityName = source.getSource() == null ? "null" : source.getSource().getEntityName();
            InfoReceptionService.InfoReceptionServiceFactory.get().dispatchWorldDeath(
                    player.getGameProfile().getId(),
                    entityName,
                    source.getType().msgId()
            );
        }

        return ActionResult.PASS;
    }

    public static void onExperienceGain(int experience, ServerPlayerEntity player) {
        InfoReceptionService.InfoReceptionServiceFactory.get().updateExperienceGained(player.getUuid(), experience);
    }

    // we need to reset timers on disconnect
    public static void onPlayerDisconnect(ServerPlayerEntity player) {
        InfoReceptionService.InfoReceptionServiceFactory.get().setIsInWater(player.getUuid(), false);
        InfoReceptionService.InfoReceptionServiceFactory.get().setIsInNether(player.getUuid(), false);
    }

    // start back up nether timer if player is created (connects) and is in the nether
    public static void onConstructPlayer(ServerWorld world, ServerPlayerEntity player) {
        if (world.getRegistryKey() == World.NETHER) {
            InfoReceptionService.InfoReceptionServiceFactory.get().setIsInNether(player.getUuid(), true);
        }
    }

    public static void onMoveToWorld(ServerWorld destination, ServerPlayerEntity player) {
        if (destination.getRegistryKey() == World.NETHER) {
            InfoReceptionService.InfoReceptionServiceFactory.get().setIsInNether(player.getUuid(), true);
        }

        if (destination.getRegistryKey() == World.OVERWORLD) {
            InfoReceptionService.InfoReceptionServiceFactory.get().setIsInNether(player.getUuid(), false);
        }
    }
}

File: src/main/kotlin/grojdg/harderstats/events/ServerPlayerEntityEvents.kt
package grojdg.harderstats.events

import net.fabricmc.fabric.api.event.Event
import net.fabricmc.fabric.api.event.EventFactory
import net.minecraft.entity.damage.DamageSource
import net.minecraft.server.network.ServerPlayerEntity
import net.minecraft.server.world.ServerWorld

object ServerPlayerEntityEvents {
    fun interface AfterDamage {
        fun interact(source: DamageSource, amount: Float, player: ServerPlayerEntity)
    }

    val AFTER_DAMAGE: Event<AfterDamage> = EventFactory.createArrayBacked(AfterDamage::class.java
    ) { listeners ->
        AfterDamage { source, amount, player ->
            for (listener in listeners) {
                listener.interact(source, amount, player)
            }
        }
    }

    fun interface AfterDeath {
        fun interact(source: DamageSource, player: ServerPlayerEntity)
    }

    val AFTER_DEATH: Event<AfterDeath> = EventFactory.createArrayBacked(AfterDeath::class.java
    ) { listeners ->
        AfterDeath { source, player ->
            for (listener in listeners) {
                listener.interact(source, player)
            }
        }
    }

    fun interface AfterExperienceGain {
        fun interact(experience: Int, player: ServerPlayerEntity)
    }

    val AFTER_EXPERIENCE_GAIN: Event<AfterExperienceGain> = EventFactory.createArrayBacked(AfterExperienceGain::class.java
    ) { listeners ->
        AfterExperienceGain { experience, player ->
            for (listener in listeners) {
                listener.interact(experience, player)
            }
        }
    }

    fun interface AfterPlayerDisconnect {
        fun interact(player: ServerPlayerEntity?)
    }

    val AFTER_PLAYER_DISCONNECT: Event<AfterPlayerDisconnect> = EventFactory.createArrayBacked(AfterPlayerDisconnect::class.java
    ) { listeners ->
        AfterPlayerDisconnect { player ->
            for (listener in listeners) {
                listener.interact(player)
            }
        }
    }

    fun interface OnMoveToWorld {
        fun interact(destination: ServerWorld, player: ServerPlayerEntity)
    }

    val ON_MOVE_TO_WORLD: Event<OnMoveToWorld> = EventFactory.createArrayBacked(OnMoveToWorld::class.java
    ) { listeners ->
        OnMoveToWorld { world, player ->
            for (listener in listeners) {
                listener.interact(world, player)
            }
        }
    }

    fun interface OnConstructPlayer {
        fun interact(world: ServerWorld, player: ServerPlayerEntity)
    }

    val ON_CONSTRUCT_PLAYER: Event<OnConstructPlayer> = EventFactory.createArrayBacked(OnConstructPlayer::class.java
    ) { listeners ->
        OnConstructPlayer { world, player ->
            for (listener in listeners) {
                listener.interact(world, player)
            }
        }
    }
}